### PR TITLE
Chore/printer manager js clean up

### PR DESCRIPTION
**Pull request by NotExpectedYet**
Initial clean up is done. 

- [x] Reactivate the printer specific functions
- [x] Reactivate the printer manager functions
- [x] Fix the enable,disable,uninstall functions after merging
- [x] Make sure all buttons call latest OP data if required.

### CODE REVIEW

**Comment by NotExpectedYet** [comment]
Functions have all been moved and cleaned. Some main functions like data updating have been improved but the rest have been left as were.

File: client_src/js/dashboard/dashboard.worker.js [hunk 1]
```diff
@@ -1,103 +0,0 @@
...
-    if (worker === null) {
-      console.log("Screen resumed... opening web worker...");
-      createWebWorker();
-    }
-  }
-}
-
-function createWebWorker() {
-  worker = new Worker(workerResource);
-  worker.onmessage = async function (event) {
```
Comment (davidzwa) [inline client_src/js/dashboard/dashboard.worker.js:25]: I'd like to revert this change. The code was neatly split up in order to see what is asynchronous through SSE and not the same as the rest of the normal sync dashboard code.
I couldve named it dashboard-client.service.js and put it in a backend client service folder or similar to make that idea clearer

File: client_src/js/dashboard.runner.js [hunk 2]
```diff
@@ -67,7 +68,68 @@ async function initNewGraphs() {
   await ChartsManager.renderPrintCompletionByDay(printCompletionByDay);
 }
 
-loadClientSSEWorker();
+const workerURL = "/dashboardInfo/get/";
+
+async function workerEventFunction(data) {
```
Comment (davidzwa) [inline client_src/js/dashboard.runner.js:73]: I'd like to keep this stuff split from the dashboard entry file, see my review comment in the worker file. I could've named this file `dashboard.entry.js`. The intention is that it binds everything together but little code is present.

File: client_src/js/lib/modules/Printers/actionButtons.js [hunk 3]
```diff
@@ -75,15 +75,15 @@ function printerQuickDisconnected(id) {
...
 
-async function init(printer, element) {
+function init(printer, element) {
   document.getElementById(element).innerHTML = `
-    ${printerControlBtn(printer._id)}  
-    ${printerWebBtn(printer._id, printer.printerURL)}  
-    ${printerReSyncBtn(printer._id)}  
-    ${printerQuickConnect(printer._id)}  
-    ${powerBtnHolder(printer._id)}  
+      ${printerControlBtn(printer._id)}
```
Comment (davidzwa) [inline client_src/js/lib/modules/Printers/actionButtons.js:80]: Indentation is off

File: client_src/js/lib/modules/Printers/actionButtons.js [hunk 4]
```diff
@@ -75,15 +75,15 @@ function printerQuickDisconnected(id) {
   connectBtn.title = "Press to connect your printer!";
 }
 
-async function init(printer, element) {
+function init(printer, element) {
```
Comment (davidzwa) [inline client_src/js/lib/modules/Printers/actionButtons.js:78]: Nice

File: client_src/js/lib/octoprint.js [hunk 5]
```diff
@@ -56,19 +56,19 @@ export default class OctoPrintClient {
...
-      headers: {
-        "Content-Type": "application/json",
-        "X-Api-Key": printer.apikey
-      },
-      body: JSON.stringify(data)
-    }).catch((e) => {
-      console.log(e);
-    });
-  }
+  // static move(printer, item, data) {
```
Comment (davidzwa) [inline client_src/js/lib/octoprint.js:59]: Please remove

File: client_src/js/printerManager/printer-constructor.js [hunk 6]
```diff
@@ -0,0 +1,323 @@
+import Validate from "../lib/functions/validate";
+import UI from "../lib/functions/ui";
+import OctoFarmClient from "../lib/octofarm_client.js";
+
+let newPrintersIndex = 0;
+
+const removeLine = function (element) {
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:7]: UI code mixed with data/logic

File: client_src/js/printerManager/printer-constructor.js [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+      UI.createAlert(
+        "error",
+        "Something went wrong saving your printer, please check the logs",
+        3000,
+        "clicked"
+      );
+    }
+  }
+
+  build() {
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:320]: Function name does not make sense

File: client_src/js/printerManager/printer-constructor.js [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+      showFahrenheitAlso: false
+    };
+    this.printerURL = printerURL;
+    this.camURL = camURL;
+    this.apikey = apikey;
+    this.group = group;
+  }
+}
+export class PrintersManagement {
+  constructor(printerURL, camURL, apikey, group, name) {
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:26]: I see a constructor for a single printer's properties in a manager pattern... pls dont

File: client_src/js/printerManager/printer-constructor.js [hunk 9]
```diff
@@ -0,0 +1,323 @@
+import Validate from "../lib/functions/validate";
+import UI from "../lib/functions/ui";
+import OctoFarmClient from "../lib/octofarm_client.js";
+
+let newPrintersIndex = 0;
+
+const removeLine = function (element) {
+  element.remove();
+};
+class Printer {
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:10]: - Split out from this file
- Nice idea but it would be nice if the printer could be serialized easily by having a `serialize` function. This way the object's data can easily be sent back to backend
- File name would be `printer.manager.js` if we'd be managing multiple printers here.

File: client_src/js/printerManager/printer-constructor.js [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+  static addPrinter(newPrinter) {
+    // Insert Blank Row at top of printer list
+    if (
+      document.getElementById("printerNewTable").classList.contains("d-none")
+    ) {
+      document.getElementById("printerNewTable").classList.remove("d-none");
+    }
+
+    if (typeof newPrinter !== "undefined") {
+      document.getElementById("printerNewList").insertAdjacentHTML(
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:39]: Is this a UI manager or a data manager?

File: client_src/js/printerManager/printer-constructor.js [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+            </button></td>
+
+    </tr>
+  `
+      );
+    } else {
+      document.getElementById("printerNewList").insertAdjacentHTML(
+        "beforebegin",
+        `
+        <tr id="newPrinterCard-${newPrintersIndex}">
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:72]: UI or data?

File: client_src/js/printerManager/printer-constructor.js [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+          3000
+        );
+      }
+    };
+  }
+
+  static async deletePrinter(deletedPrinters) {
+    if (deletedPrinters.length > 0) {
+      try {
+        const printersToRemove = await OctoFarmClient.post(
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:177]: Data vs UI. This is data

File: client_src/js/printerManager/printer-constructor.js [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+          "printers/remove",
+          deletedPrinters
+        );
+        // Should help with SSE event re-building the deleted printer causing a refresh to be needed to actually clear from UI
+        await UI.delay(5000);
```
Comment (davidzwa) [inline client_src/js/printerManager/printer-constructor.js:182]: We should avoid this at **all cost**. If we have problems with delayed data then fix the data stream, not the original point where it's requested

This is a huge red flag especially 3000 sec

File: client_src/js/printerManager/templates/gcode-scripts-table.template.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
+export function customGcodeScriptTemplate(scripts, scriptLines) {
```
Comment (davidzwa) [inline client_src/js/printerManager/templates/gcode-scripts-table.template.js:1]: Love this idea!

File: client_src/js/utils/time.util.js [hunk 15]
```diff
@@ -21,3 +21,11 @@ export function getLastThirtyDaysText() {
 
   return lastThirtyDaysText;
 }
+
+/**
+ * Format a date string to human readable date.
+ * @returns {String} Human readable date string
+ */
+export function formatDateStringHumanReadable(date) {
```
Comment (davidzwa) [inline client_src/js/utils/time.util.js:29]: Shorten formatter as this function name is longer than the normal call

File: client_src/package.json [hunk 16]
```diff
@@ -59,7 +59,6 @@
     "jplist-es6": "^1.2.0",
     "jquery": "^3.5.1",
     "lodash": "^4.17.21",
-    "luxon": "^1.26.0",
```
Comment (davidzwa) [inline client_src/package.json:62]: Why? Thought we wanted to standardize date formatting

File: client_src/js/printerManager/printer-constructor.js (hunk 13 above)
Comment (NotExpectedYet) [reply client_src/js/printerManager/printer-constructor.js:182]: 5 seconds not 3000

also outside of the scope of this PR

File: client_src/js/printerManager/printer-constructor.js (hunk 8 above)
Comment (NotExpectedYet) [reply client_src/js/printerManager/printer-constructor.js:26]: This was mainly aimed at splitting the god awful printerManagement file down not at refactoring function as mentioned. Outside of scope of PR, please keep within scope.

File: client_src/js/printerManager/printer-constructor.js (hunk 13 above)
Comment (davidzwa) [reply client_src/js/printerManager/printer-constructor.js:182]: I meant to write 3000ms. Thats still wrong on my side! 😄 

5 seconds is a long time and there's no cancellation. All I can see it's introduced in this PR, so I feel the obligation to review it here as well.

Comment (NotExpectedYet) [reply client_src/js/printerManager/printer-constructor.js:182]: It's a patch work to the current method of SSE events. When the delete function is fired, this function get's the acceptance response from the server side code, before the printerClean function has actually cleaned out the removed printer(s). Fixing the data stream requires a lot more work than is contained within this PR, because there's nothing within the backend to know about the printerClean function successfully running from the API endpoint. The delay doesn't lock up the client, and it also does nothing to the delete function been fired, it just delays the actual removal of the element and the UI notification trigger for the user.

**Review by davidzwa — CHANGES_REQUESTED** [review]
Great potential in this PR as you said it's a good way to get started.
I've really tried my best to ignore the things you don't want me to review by resolving the conversations.

I need to request some changes mainly on the dashboard runner as this is code I refactored recently and it's in my interest to keep it clean. See the 2 comments above on that.
2nd one is some commented out code, also see comment above.
3rd is an indentation, so optional.

Could you show my the changes giving a `little speed boost` as mentioned in changelog. Interested to know.

Should be a breeze to fix.

**Comment by NotExpectedYet** [comment]
> Great potential in this PR as you said it's a good way to get started.
> I've really tried my best to ignore the things you don't want me to review by resolving the conversations.
> 
> I need to request some changes mainly on the dashboard runner as this is code I refactored recently and it's in my interest to keep it clean. See the 2 comments above on that.
> 2nd one is some commented out code, also see comment above.
> 3rd is an indentation, so optional.
> 
> Could you show my the changes giving a `little speed boost` as mentioned in changelog. Interested to know.
> 
> Should be a breeze to fix. 💪🏼

The speed boost was lower CPU usage due to not having the elements constantly updating as before, and it now reacts to the first SSE message where it didn't before. So you get the initial information quicker. It's hard to quantify tbh it's more just something I noticed. Can change the note if it's mis-leading. Will sort the rest.

**Comment by NotExpectedYet** [comment]
> Great potential in this PR as you said it's a good way to get started.
> I've really tried my best to ignore the things you don't want me to review by resolving the conversations.
> 
> I need to request some changes mainly on the dashboard runner as this is code I refactored recently and it's in my interest to keep it clean. See the 2 comments above on that.
> 2nd one is some commented out code, also see comment above.
> 3rd
 is an indentation, so optional.
> 
> Could you show my the changes giving a `little speed boost` as mentioned in changelog. Interested to know.
> 
> Should be a breeze to fix. 💪🏼

Ran two tests using firefox performance monitor. Both we're started and then page refreshed until buffer full with 99 printers in the same state for each test. 

New: (avg ~45)
image


Old: (avg ~31fps) 
image

File: client_src/js/printerManager/printer-constructor.js (hunk 10 above)
Comment (NotExpectedYet) [reply client_src/js/printerManager/printer-constructor.js:39]: It's an old god awful amalgamation of both, left a note to clean up that file as it's got stuff tagged on from old usages.

File: client_src/js/printerManager/templates/gcode-scripts-table.template.js (hunk 14 above)
Comment (NotExpectedYet) [reply client_src/js/printerManager/templates/gcode-scripts-table.template.js:1]: Yeah do like this myself it's so much easier and re-usable across the UI

File: client_src/js/utils/time.util.js (hunk 15 above)
Comment (NotExpectedYet) [reply client_src/js/utils/time.util.js:29]: I wrote the wrapper as I'd originally planned to use the formatter but decided that should probably be done in another PR. I'll revert this as it's only used in printer connection log. As you say is unneeded.

**Review by davidzwa — APPROVED** [review]
Page feels faster, looks better. Love it. The client code changes: no idea === LGTM
Ofc we should try to work on better sync with the server if delays are required to fix things.

You know me, I hate waiting XD

**Comment by davidzwa** [comment]
image
Undefined when not OctoPi
Printer disconnected => unknown printer shown

**Comment by NotExpectedYet** [comment]
> Page feels faster, looks better. Love it. The client code changes: no idea === LGTM
> Ofc we should try to work on better sync with the server if delays are required to fix things.

> You know me, I hate waiting XD

I know XD 

What I wanted to do was have the client notice when a ID disappears from the printerInformation it removes it itself. Dunno happy for thoughts on how to handle it, was just something I noticed whilst testing the delete function and forgot it needed patching from ages ago. It totally depends how in sync your click is with printer clean. 

> image
> Undefined when not OctoPi
> Printer disconnected => unknown printer shown

That needs putting in the database imo, OctoPrint doesn't send it when the printer is disconnected. It was lacking from me originally adding the version. I'll do a ticket for it if you haven't already.
